src/Tmds.Ssh/SftpClient.cs

```diff
@@ -0,0 +1,205 @@
+            payload.Remove(9);
+            while (ReadSftpPacket(payload, out SftpPacketType type, out uint requestId, out ReadOnlySequence<byte> fields, out uint consumed))
+            {
+                if (_operations.TryGetValue(requestId, out SftpOperation operation))
+                {
+                    await operation.HandleResponse(type, fields, this);
+                }
+                payload.Remove(consumed);
+            }
+        }
+
+        private bool ReadSftpPacket(Sequence payload, out SftpPacketType type, out uint requestId, out ReadOnlySequence<byte> fields, out uint consumed)
+        {
+            if (payload.IsEmpty)
+            {
+                type = default;
+                requestId = 0;
+                fields = default;
+                consumed = 0;
+                return false;
+            }
+            /*
+                uint32           length        // The length of the entire packet, excluding the length field
+                byte             type
+                uint32           request-id
+                    ... type specific fields ...
+            */
+            var reader = new SequenceReader(payload);
+            uint length = reader.ReadUInt32();
+            type = reader.ReadSftpPacketType();
+            requestId = reader.ReadUInt32();
+            fields = payload.AsReadOnlySequence().Slice(9, length - 5);
+            consumed = 4 + length;
+            return true;
+        }
+
+        private async ValueTask<uint> ReceiveServerVersionAsync(string failureMessage, CancellationToken ct)
+        {
+            using var packet = await _context.ReceivePacketAsync(ct).ConfigureAwait(false);
+
+            return ParseSftpVersion(packet, failureMessage);
+
+            static uint ParseSftpVersion(ReadOnlyPacket packet, string failureMessage)
+            {
+                /*
+                            byte            SSH_MSG_CHANNEL_DATA
+                            uint32          recipient channel
+                            string          data
+
+                            uint32          SftpLength
+                            byte            SftpType
+                            uint32          SftpVersion
+                            string          extension-name
+                            string          extension-data
+                */
+                var reader = packet.GetReader();
+                reader.ReadMessageId(MessageId.SSH_MSG_CHANNEL_DATA);
+                reader.Skip(12);
+                reader.ReadSftpPacketType(SftpPacketType.SSH_FXP_VERSION);
+                var version = reader.ReadUInt32();
+                return version;
+            }
+        }
+
+        private ValueTask SftpInitMessageAsync(uint version, CancellationToken ct)
+        {
+            /*
+                    uint32 	length
+                    byte 	type
+                    uint32 	version
+            */
+
+            var memory = new Memory<byte>(new byte[9]);
```

Comment:
Can you implement this similar to the methods in `ChannelContextSendMessageExtensions.cs` (using `context.RentPacket();` etc)?

Reply:
That's what was there before, but in the comment above you said to use the SendChannelDataMessageAsync, that takes Memory<T>, and I am not sure how to create Memory<T> from the packet to use the method :thinking:

Reply:
I meant ChannelContext.SendChannelData~~Message~~Async which takes a Packet arg.

Reply:
Oh ye, that makes sense